source/common/buffer/watermark_buffer.h

```diff
@@ -72,15 +75,128 @@ class WatermarkBuffer : public OwnedImpl {
+    ASSERT(!reset_handler_.has_value());
+  }
+
+  // Make not copyable
+  BufferMemoryAccountImpl(const BufferMemoryAccountImpl&) = delete;
+  BufferMemoryAccountImpl& operator=(const BufferMemoryAccountImpl&) = delete;
+
+  // Make not movable.
+  BufferMemoryAccountImpl(BufferMemoryAccountImpl&&) = delete;
+  BufferMemoryAccountImpl& operator=(BufferMemoryAccountImpl&&) = delete;
+
+  uint64_t balance() const { return buffer_memory_allocated_; }
+  void charge(uint64_t amount) override;
+  void credit(uint64_t amount) override;
+
+  // Clear the associated downstream, preparing the account to be destroyed.
+  // This is idempotent.
+  void clearDownstream() override;
+
+  void resetDownstream() override {
+    if (reset_handler_.has_value()) {
+      reset_handler_->resetStream(Http::StreamResetReason::OverloadManager);
+    }
+  }
+
+  // The number of memory classes the Account expects to exists. See
+  // *WatermarkBufferFactory* for details on the memory classes.
+  static constexpr uint32_t NUM_MEMORY_CLASSES_ = 8;
+
+private:
+  BufferMemoryAccountImpl(WatermarkBufferFactory* factory, Http::StreamResetHandler& reset_handler)
+      : factory_(factory), reset_handler_(reset_handler) {}
+
+  // Returns the class index based off of the buffer_memory_allocated_
+  // This can differ with current_bucket_idx_ if buffer_memory_allocated_ was
+  // just modified.
+  // The class indexes returned are based on buckets of powers of two, if the
+  // account is above a minimum threshold. Returned class index range is [-1,
+  // NUM_MEMORY_CLASSES_).
+  int balanceToClassIndex();
+  void updateAccountClass();
+
+  uint64_t buffer_memory_allocated_ = 0;
+  // Current bucket index where the account is being tracked in.
+  int current_bucket_idx_ = -1;
+
+  WatermarkBufferFactory* factory_ = nullptr;
+
+  OptRef<Http::StreamResetHandler> reset_handler_;
+  // Keep a copy of the shared_ptr pointing to this account. We opted to go this
+  // route rather than enable_shared_from_this to avoid wasteful atomic
+  // operations e.g. when updating the tracking of the account.
+  // This is set through the createAccount static method which is the only way to
+  // instantiate an instance of this class. This should is cleared when
+  // unregistering from the factory.
+  BufferMemoryAccountSharedPtr shared_this_ = nullptr;
+};
+
+/**
+ * The WatermarkBufferFactory creates *WatermarkBuffer*s and
+ * *BufferMemoryAccountImpl* that can be used to bind to the created buffers
+ * from a given downstream (and corresponding upstream, if one exists). The
+ * accounts can then be used to reset the underlying stream.
+ *
+ * Any account produced by this factory might tracked by the factory using the
+ * following scheme:
+ *
+ * 1) Is the account balance >= 1MB? If not don't track.
+ * 2) For all accounts above the minimum threshold for tracking, put the account
+ *    into one of the *BufferMemoryAccountImpl::NUM_MEMORY_CLASSES_* buckets. As
```

Comment:
Can you explain why classes exist to begin with?

Reply:
Done.